EE: Dont deserialize a value `V` when writing a trie or when reading a trie if the key was not found

alsrdn opened the pull request:
Make sure we don't deserialize a value `V` when writing a trie.
Also put in place tests to verify this doesn't happen and that reading a trie deserializes `V` only once if the key is found or 0 times if the key is not found. The tests were extended by using the existing counter that is used to track the call to `from_bytes` for `V`. The tracker can be used now with any operation, not only `delete`.

Fixes: https://github.com/casper-network/casper-node/issues/3850 and https://github.com/casper-network/casper-node/issues/3851

alsrdn committed:
ee/tests: extend tests to check V deserialization on read/write

Extend testing logic to check if reading a value from a trie
deserializes it once if the key is found, or zero times if the key does
not exist.
Extend testing logic to check if writing a value doesn't deserialize it
during a write operation.
Extended the existing counter that tracks calls to `from_bytes` for `V`
to work with `read` and `write` operations.

Signed-off-by: Alexandru Sardan <alexandru@casperlabs.io>

alsrdn committed:
ee/trie_store: don't deserialize V during write operation

Avoid deserializing `V` during a `write` operation to the trie_store.

Signed-off-by: Alexandru Sardan <alexandru@casperlabs.io>

alsrdn committed:
ee/trie_store: remove unused `scan` operation

Remove unused `scan` operation in favor of `scan_raw`. Adjust the tests
to use `scan_raw` instead.

Signed-off-by: Alexandru Sardan <alexandru@casperlabs.io>

alsrdn committed:
Merge branch 'feat-1.6' into 3850-3851-read-write-dont-deserialize-V

alsrdn committed:
ee/trie_store: use NonDeserializingStore in `scan_raw`

Signed-off-by: Alexandru Sardan <alexandru@casperlabs.io>

EdHastingsCasperAssociation commented on execution_engine/src/storage/trie_store/operations/tests/mod.rs:
```diff
@@ -15,6 +15,8 @@ use std::{
 };
 
+use backtrace::Backtrace;
+use libc::c_void;
```
can I get some info about the choice to attempt backtracing etc for these tests?

alsrdn replied:
We wanted to track the number of times `from_bytes` was called when deserializing a value. One way to do this is to track the callers of `from_bytes` and determine if it's one of the functions that we are trying to restrict. Now we use a different strategy to achieve the same thing (with `NonDeserializingStore`, `OnceDeserializingStore` and `PanickingFromBytes`) so `backtrace` is not needed anymore.

rafal-ch commented on execution_engine/src/storage/trie_store/operations/mod.rs:
```diff
@@ -925,52 +890,67 @@ where
-                    rehash(new_leaf, parents)?
+                LazyTrieLeaf::Left(leaf_bytes) => {
+                    let trie_tag = trie::lazy_trie_tag(leaf_bytes.as_slice());
+                    assert_eq!(
+                        trie_tag,
+                        Some(TrieTag::Leaf),
+                        "Unexpected trie variant found instead of a `TrieTag::Leaf`"
+                    );
```
Are we sure we want to panic inside `write()` in case of mismatch?

mpapierski replied:
Currently, it's guaranteed by lazy trie deserializer that the panic wouldn't happen - we have ticket #3994 to address this by using proper enum rather than Either

rafal-ch commented on execution_engine/src/storage/trie_store/operations/tests/mod.rs:
```diff
@@ -129,13 +132,27 @@ impl ToBytes for TestValue {
     }
 }
 
+// Determine if a there exists a caller in the backtrace that matches any of the specified symbols
```
Nit:
```suggestion
// Determine if there exists a caller in the backtrace that matches any of the specified symbols
```

alsrdn replied:
Fixed in https://github.com/casper-network/casper-node/pull/3984/commits/60be67bc699821fc9b2f664f05a8ac6db2b057d2

alsrdn committed:
ee/trie_store/tests: use `PanickingFromBytes` for values when writing

Using `PanickingFromBytes` in tests ensures the `write` operation will
not deserialize a value.

Signed-off-by: Alexandru Sardan <alexandru@casperlabs.io>

alsrdn committed:
ee/trie_store/operation: use `PanickingFromBytes` wrapper

Use `PanickingFromBytes` wrapper for values `V` in tests to ensure a
value is not deserialized by `delete` operations.
Also use the `NonDeserializingStore` wrapper in the operation themselves
to strenghten this guarantee for debug runs.

Signed-off-by: Alexandru Sardan <alexandru@casperlabs.io>

alsrdn committed:
ee/trie_store/operations: remove `from_bytes` caller tracking

Remove the tracking of callers to `from_bytes` when using `write` and
`delete` operations. `V` is guaranteed not to be deserialized in these
cases due to the usage of `PanickingFromBytes<V>` and
`NonDeserializingStore`.

Signed-off-by: Alexandru Sardan <alexandru@casperlabs.io>

alsrdn committed:
ee/trie_store: add `OnceDeserializingStore` wrapper

Add `OnceDeserializingStore` wrapper used to ensure that the read
operation does not deserialize a value more than once.

Signed-off-by: Alexandru Sardan <alexandru@casperlabs.io>

alsrdn committed:
ee/trie_store: fix build and clippy issues

Signed-off-by: Alexandru Sardan <alexandru@casperlabs.io>

alsrdn committed:
ee/trie_store/tests: remove caller tracking for deserializing V

Now we use `OnceDeserializingStore` to ensure that the `read` operation
does not deserialize `V`.
Remove the usage of the caller tracking that used `backtrace` since it's
not needed anymore.

Signed-off-by: Alexandru Sardan <alexandru@casperlabs.io>

rafal-ch reviewed (approved):
LGTM :+1:

alsrdn commented:
bors merge
